packages/xsnap-lockdown/README.md

```diff
@@ -0,0 +1,85 @@
+# xsnap-lockdown
+
+This package provides the bundles necessary to prepare an Endo (aka SES: Secure ECMAScript) environment within an `xsnap` worker process.
+
+The basic xsnap package (@agoric/xsnap) provides two components. The first is the `xsnap` executable, which embeds the XS JavaScript engine inside a driver program that listens for commands on a socket. The second is a library which can launch that program as a child process, and sends/receives messages over the socket. The `@agoric/xsnap` API lets you perform three basic operations on the worker: evaluate a string of code, deliver a string to a globally-registered handler function, and instruct the XS engine to write out a heap snapshot.
+
+However this is not quite sufficient for use as a vat host, which needs a SES environment which can accept vat deliveries and issue syscalls. To build something more suitable, we need additional layers:
+
+* We must evaluate a "lockdown" bundle within the worker, transforming it from plain (unsecured) JS to our preferred Endo environment. This tames the global constructors, adds the `Compartment` constructor, and removes ambient authority. It also creates a `console` object around the basic XS `print` function.
+* Then, we need to install a "supervisor" (i.e. evaluate the supervisor bundle), which hooks into the globally-registered handler function to accept delivery messages. The supervisor includes the "liveslots" code, which constructs an object-capability environment with messages routed through syscalls, as well as virtual/durable object support (through the vat-data package).
+* Finally, we'll install the actual vat bundle and run its `buildRootObject()` function.
+
+This package provides the "lockdown" bundle, which incorporates Endo and the console constructor. The bundle is generated at build time, so the contents are fixed for any given version of the `@agoric/xsnap-lockdown` package. The exported API includes a function which returns the lockdown bundle contents, as well as other helper bundles for debugging purposes.
```

Comment:
_definitely covered later_

Reply:
The bundle is the same for everybody because everybody gets it from the NPM published tarball. If you rebuild the bundle locally, the potential sources of differences from the one that build (and publish to NPM) are:

* the exact versions of Endo and other incorporated dependency code that were selected by your dependency solver
  * this package's `package.json` provides version constraints for those dependencies, as usual
  * `yarn.lock` provides a stronger constraint, when checked in, however:
  * as long as this package lives in a repo with other packages, the shared `yarn.lock` will override
  * so some other package wanting a newer-yet-still-meets-my-constraints version will win
  * also `lerna version` will cheerfully update our package.json file, probably in surprising ways
  * if/when we move this package into a separate repo, it will probably have its own `yarn.lock`, which will help
* the exact behavior of the Endo `bundleSource` tool, which is slightly more likely to remain the same from one version to the next

Absolute path names should *not* get embedded in the generated bundle: @kriskowal fixed that a while ago. The pathname you get from `import { bundlePaths } from '@agoric/xsnap-lockdown'` will vary depend upon where you've checked out your tree, but that's not part of the bundle.

The hash is meant as a failsafe. We'll see how well it works, when I try to build the higher-level `@agoric/swingset-worker-xs-v1` package, and use it in anger.